### DISCUSSION TITLE
Capped (but not constrained) max_level of reservoir

### DISCUSSION THREAD

**Discussion by evertheylen** [google/or-tools#3516] [CP-SAT questions] [answered]
Hello,

I'm trying to model the scheduling of tasks, where tasks use a certain resource (e.g. money) that is replenished daily. Every task has a certain cost that has to be paid when it is scheduled. However, there should be limits to how much money can be held without spending it.

It seems like reservoirs are the way to go:
- I can set `min_level`, `max_level` as desired
- I can model the task costs as negative demands, at a variable time
- I can model the replenishing of the money as positive demands, on each time from 0 until my time horizon (= `maximum_duration`)

There is one problem: replenishing the reservoir above `max_level` should not be considered wrong. Instead it should just ignore that and consider the reservoir capped at `max_level` (one could call it overflow maybe?). To clarify further, if the solver has (hypothetically) correctly scheduled all tasks but my time horizon is too long, there are days at the end during which it tries to add money beyond `max_level`, ruining an otherwise perfectly good solution.

I thought I could make this work by using the `ReservoirConstraintWithActive` and having an unconstrained variable when I add the positive demand. However, this makes my runtime and memory requirements explode. The alternative is to leave the reservoir unbounded, but this is does not (always) give the same results.

I made a small example to clarify. In here, `LIMIT_RESERVOIR=True` represents the correct by super slow version, and `LIMIT_RESERVOIRS=False` represents the wrong but much faster solution. In this particular example, the solutions found are equivalent. However, the first version runs 50 times slower. **How can I get the behaviour of a "capped" reservoir without the extreme time/memory requirements?**

```python
from ortools.sat.python import cp_model
import time

task_costs = [1, 2, 3, 4] * 5
daily_added_money = 4
maximum_duration = (sum(task_costs) // daily_added_money) + 2  # Some high enough upper bound

def define_problem(LIMIT_RESERVOIR):
  model = cp_model.CpModel()
  task_start_vars = [model.NewIntVar(0, maximum_duration, f'start_{i}') for i, t in enumerate(task_costs)]

  demands = []
  actives = []
  times = []
  
  for i, task_cost in enumerate(task_costs):
    demands.append(-task_cost)
    times.append(task_start_vars[i])
    actives.append(model.NewConstant(True))

  for d in range(maximum_duration):
    demands.append(+daily_added_money)
    times.append(model.NewConstant(d))
    if LIMIT_RESERVOIR:
      # After all tasks are done, make sure we don't go over our limit simply by running out
      # of tasks. We do this by making the extra added money optional (unconstrained variable)
      actives.append(model.NewBoolVar(f'add_money_{d}'))
    else:
      # There is no limit, so we can reduce the variables by always adding money
      actives.append(model.NewConstant(True))

  reservoir = model.AddReservoirConstraintWithActive(
    times, demands, actives,
    min_level=0,
    max_level=daily_added_money if LIMIT_RESERVOIR else 99999999,
  )

  all_done = model.NewIntVar(0, maximum_duration, f'all_done')
  model.AddMaxEquality(all_done, task_start_vars)
  model.Minimize(all_done*1000 + sum(task_start_vars))

  return model, task_start_vars


def solve(model, task_start_vars):
  start = time.process_time()
  try:
    solver = cp_model.CpSolver()
    status = solver.Solve(model)

    if status == cp_model.OPTIMAL or status == cp_model.FEASIBLE:
      print(f'\nMaximum of objective function: {solver.ObjectiveValue()}\n')
    else:
      print('No solution found.')
      return
  finally:
    end = time.process_time()
    print(f"---> took {end - start:.3f}s")

  schedule = [(i, task_cost, solver.Value(task_start)) for i, (task_cost, task_start) in enumerate(zip(task_costs, task_start_vars))]
  schedule.sort(key=lambda x: x[2])
  for i, cost, start in schedule:
    print(f"At time {start}, run task {i} (cost {cost})")


print("-------------- WITH RESERVOIR LIMITS ---------------")
solve(*define_problem(LIMIT_RESERVOIR=True))

print("\n-------------- WITHOUT RESERVOIR LIMITS ---------------")
solve(*define_problem(LIMIT_RESERVOIR=False))

```

**Comment by lperron** [comment, marked as answer]
I think you need to change the model.

Imagine the refill happens periodically. Then the model is actually in which period consumption happens.
So you just need to have one var for the level at the start of each period, one bool var per period and per demand to indicates if the consumptions will happen in that period, and one variable for the refill at the start of each period. 

Then you just need to write the conservation of matter

    level_at_start_of_period[p] - sum(consumptions_in_that_period) + refill_at_start_of_period[p + 1] == level_at_start_of_period[p + 1]
